### PR TITLE
[Compression] pruning speedup replace transformer attention layers

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
An external extension to speedup huggingface transformers attention by using built-in `prune_heads`.

TODO:
* [ ] - doc for how to customize replacor
* [ ] - test

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: examples/model_compress/pruning/speedup_attention.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    'op_types': ['Linear'],
+    'op_partial_names': ['bert.encoder.layer.{}.attention.self'.format(i) for i in range(12)],
+    'sparsity': 0.98
+}]
+
+pruner = L1NormPruner(model, config_list)
+_, masks = pruner.compress()
+pruner._unwrap_model()
+
+replacor = TransformersAttentionReplacor(model)
```
Comment (QuanluZhang) [inline examples/model_compress/pruning/speedup_attention.py:32]: replacer

File: nni/compression/pytorch/utils/external/huggingface.py [hunk 2]
```diff
@@ -21,37 +21,11 @@
 else:
     TRANSFORMERS_INSTALLED = True
 
-from nni.compression.pytorch.utils.attr import get_nested_attr
+from nni.algorithms.compression.v2.pytorch.utils.attr import get_nested_attr
```
Comment (QuanluZhang) [inline nni/compression/pytorch/utils/external/huggingface.py:24]: we removed "v2", right?

File: examples/model_compress/pruning/speedup_attention.py (hunk 1 above)
Comment (J-shang) [reply examples/model_compress/pruning/speedup_attention.py]: fixed

File: nni/compression/pytorch/utils/external/huggingface.py (hunk 2 above)
Comment (J-shang) [reply nni/compression/pytorch/utils/external/huggingface.py]: fixed

File: nni/compression/pytorch/speedup/compressor.py [hunk 3]
```diff
@@ -43,23 +45,18 @@ class ModelSpeedup:
...
+        a ``Module`` that contains a ``Module`` as a hyper-module, then replacer is used to replace the hyper-module.
+        The difference between the replacer and replace function is that replacer can perform more efficient replacements
+        to hyper-module, and replace function is used to replace leaf-module.
+        In ``ModelSpeedup.compress``, replacers are first to be called to replace the hyper-modules before
+        replacing all leaf-modules by replace functions.
     """
 
     def __init__(self, model, dummy_input, masks_file, map_location=None,
-                 batch_dim=0, confidence=8, customized_replace_func=None):
+                 batch_dim=0, confidence=8, customized_replacers=None, customized_replace_func=None):
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/compressor.py:59]: might be better to add "customized_repalce_func" in docstring and add "deprecated" warning in it.

**Review by QuanluZhang — APPROVED** [review]

**Comment by QuanluZhang** [comment]
resolve conflict and add doc

File: nni/compression/pytorch/speedup/replacer.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+_logger = logging.getLogger()
+
+
+class Replacer:
+    def replace_modules(self, model: torch.nn.Module, auto_inferences: Dict[str, AutoMaskInference]):
+        raise NotImplementedError()
+
+
+class BuiltinReplacer:
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/compression/pytorch/speedup/replacer.py]: OK, fixed

**Review by Louis-J — APPROVED** [review]

**Pull request merged**